horovod/tensorflow/__init__.py

```diff
@@ -237,3 +248,84 @@ def get_slot_names(self, *args, **kwargs):
+        If False users must explicitly `watch` any `Variable`s they want to
+        request gradients from.
+      device_dense:
+        Device to be used for dense tensors. Uses GPU by default
+        if Horovod was build with HOROVOD_GPU_ALLREDUCE.
+      device_sparse:
+        Device to be used for sparse tensors. Uses GPU by default
+        if Horovod was build with HOROVOD_GPU_ALLGATHER.
+      compression:
+        Compression algorithm used during allreduce to reduce the amount
+        of data sent during the each parameter update step.  Defaults to
+        not using compression.
+      sparse_as_dense:
+        Treat all sparse gradients as dense tensors.  This can help improve
+        performance and memory utilization if the original sparse gradient
+        has high density.  Defaults to false.
+    """
+
+    def __init__(self, persistent, watch_accessed_variables,
+                 tape, device_dense='', device_sparse='',
+                 compression=Compression.none, sparse_as_dense=False):
+        super(self.__class__, self).__init__(persistent, watch_accessed_variables)
+        self._tape = tape
+        self._persistent = persistent
+        self._watch_accessed_variables = watch_accessed_variables
+        self._name = "Distributed"
+        self._device_dense = device_dense
+        self._device_sparse = device_sparse
+        self._compression = compression
+        self._sparse_as_dense = sparse_as_dense
+
+        def allreduce_grads(grads):
+            with tf.name_scope(self._name + "_Allreduce"):
+                if self._sparse_as_dense:
+                    grads = [tf.convert_to_tensor(grad)
+                             if grad is not None and isinstance(grad, tf.IndexedSlices)
+                             else grad for grad in grads]
+                return [allreduce(grad,
+                                  device_dense=self._device_dense,
+                                  device_sparse=self._device_sparse,
+                                  compression=self._compression)
+                        if grad is not None else grad
+                        for grad in grads]
+
+        self._allreduce_grads = tf.contrib.eager.defun(allreduce_grads)
+
+    def gradient(self, target, sources, output_gradients=None):
+        grads = super(self.__class__, self).gradient(target, sources, output_gradients)
+        if size() > 1:
+            avg_grads = self._allreduce_grads(grads)
+            return avg_grads
+        else:
+            return gradients
+
+
+def DistributedGradientTape(gradtape, device_dense='', device_sparse='', compression=Compression.none,
+                            sparse_as_dense=False):
+    cls = type(gradtape.__class__.__name__, (gradtape.__class__,),
+               dict(_DistributedGradientTape.__dict__))
+    return cls(gradtape._persistent, gradtape._watch_accessed_variables,
+               gradtape._tape,
+               device_dense='', device_sparse='',
```

Comment:
Should pass `device_dense` and etc. correctly.

Reply:
i updated it.